src/common/gpio_motor.rs

```diff
@@ -437,3 +459,146 @@ where
     }
 }
+
+/// Represents a motor with an A and B pin. When moving forwards,
+/// a PWM signal is sent through the A pin and the B pin is set to high,
+/// vice versa for moving backwards. Note: If the dir_flip attribute is set to
+/// true, this functionality is reversed
+pub(crate) struct AbMotor<B> {
+    board: B,
+    a_pin: i32,
+    b_pin: i32,
+    max_rpm: f64,
+    dir_flip: bool,
+    is_on: bool,
+    pwm_pin: i32,
+}
+
+impl<B> AbMotor<B>
+where
+    B: Board,
+{
+    pub(crate) fn new(
+        a_pin: i32,
+        b_pin: i32,
+        max_rpm: f64,
+        dir_flip: bool,
+        board: B,
+    ) -> anyhow::Result<Self> {
+        let mut res = Self {
+            board,
+            a_pin,
+            b_pin,
+            max_rpm,
+            dir_flip,
+            is_on: false,
+            pwm_pin: a_pin,
+        };
+        // we start with this because we want to reserve a timer and PWM channel early
+        // for boards where these are a limited resource
+        res.board.set_pwm_frequency(a_pin, MOTOR_PWM_FREQUENCY)?;
+        res.board.set_pwm_duty(a_pin, 0.0)?;
+        Ok(res)
+    }
+
+    pub(crate) fn from_config(cfg: ConfigType, board: BoardType) -> anyhow::Result<MotorType> {
+        if let Ok(pins) = cfg.get_attribute::<MotorPinsConfig>("pins") {
+            if pins.a.is_some() && pins.b.is_some() {
+                let a_pin = pins.a.unwrap();
+                let b_pin = pins.b.unwrap();
+                let max_rpm: f64 = cfg.get_attribute::<f64>("max_rpm").unwrap_or(100.0);
+                let dir_flip: bool = cfg.get_attribute::<bool>("dir_flip").unwrap_or_default();
+                return Ok(Arc::new(Mutex::new(AbMotor::new(
+                    a_pin, b_pin, max_rpm, dir_flip, board,
+                )?)));
+            }
+        }
+        Err(anyhow::anyhow!("cannot build motor"))
+    }
+
+    fn turn_off(&mut self) -> anyhow::Result<()> {
+        self.board.set_pwm_duty(self.pwm_pin, 0.0)?;
+        self.board.set_gpio_pin_level(self.a_pin, false)?;
+        self.board.set_gpio_pin_level(self.b_pin, false)?;
+        self.is_on = false;
+        Ok(())
+    }
+}
+
+impl<B> Motor for AbMotor<B>
+where
+    B: Board,
+{
+    fn set_power(&mut self, pct: f64) -> anyhow::Result<()> {
+        if !(-1.0..=1.0).contains(&pct) {
+            anyhow::bail!("power outside limit, must be between -1.0 and 1.0")
+        }
+        if pct.abs() <= 0.001 {
+            return self.turn_off();
+        }
+        let (pwm_pin, high_pin) = if (pct >= 0.001) != self.dir_flip {
+            (self.b_pin, self.a_pin)
+        } else {
+            (self.a_pin, self.b_pin)
+        };
+        if pwm_pin != self.pwm_pin {
+            self.board.set_pwm_frequency(pwm_pin, MOTOR_PWM_FREQUENCY)?;
+            self.board.set_pwm_frequency(self.pwm_pin, 0)?;
+        }
+        self.pwm_pin = pwm_pin;
+        self.board.set_gpio_pin_level(high_pin, true)?;
+        let power_pct = 1.0 - pct.abs();
```

Comment:
why do we need to do that?

Reply:
ok, I based this off rdk's example. I now see this was a mistake because RDK writes a single implementation for all three types of motors. I'll remove this and reverse the condition below